### ISSUE TITLE
new stats API `scopeFromStatNameWithTags`

### ISSUE THREAD

**Issue by wbpcode** [envoyproxy/envoy#20220]
*Title*: new stats API `scopeFromStatNameWithTags`

*Description*:

Now the APIs `*FromStatNameWithTags` are provided to create counter, gauge, and histogram. Can we add the new API of `scopeFromStatNameWithTags`. Then we can create scope with some static tags. Any new metrics created from this scope will has related tags directly. 🤔

**Comment by wbpcode** [comment]
cc @jmarantz

**Comment by wbpcode** [comment]
@jmarantz Thanks. 

This feature request was inspired by the #19663. In the #19663, I want create scopes based on the metadata config. Scopes with tags can be used to avoid writing complex regex extractors.

Considering this is a generic feature, I'm going to prepare a separate PR for it.

**Comment by jmarantz** [comment]
Will definitely want to get @mattklein123 blessing as well for this strategy and how you will use it.

**Comment by jmarantz** [comment]
I had another thought about this -- it seems wasteful to replicate the same tag name/value across all stats in a scope. Instead, we could have mutable tags on the scope, and you could change those in a config update without rebuilding the scope.

Then, when reading the stats, you could apply the semantics from the scope's tags. If the consumer is implemented by traversing over scopes then you'll have its tags available. This avoids memory bloat and also resetting the stats when you update the scope tags.

This will be straightforward in the context of stats sinks once https://github.com/envoyproxy/envoy/pull/19693 lands, because it makes stat rendering process scope-aware. Is that your ultimate goal, to get these new scope-wide stat stats applied across each stat during the stats-sink process, say, to Prometheus? #19693 does not actually bring scope-awareness to the Prometheus formatter but that could be a follow-up.

**Comment by wbpcode** [comment]
> it seems wasteful to replicate the same tag name/value across all stats in a scope. Instead, we could have mutable tags on the scope, and you could change those in a config update without rebuilding the scope.
> Then, when reading the stats, you could apply the semantics from the scope's tags. If the consumer is implemented by traversing over scopes then you'll have its tags available. This avoids memory bloat and also resetting the stats when you update the scope tags.

Great. I have had some similar ideas before. This method can save some memory. 🤔 

Since the stats need to be created with the full name of the stats as a unique identifier, this means that even if we don't store scope tags in the stats, we will inevitably need to rebuild all the stats under scope when scope tags are updated.

**Comment by mattklein123** [comment]
+1 this sounds good to me, especially the part about just storing tags for the entire scope and not on a per stat basis. It seems much simpler.

**Comment by wbpcode** [comment]
> I do not see why you need to rebuild the stats with the full name including the dynamically synthesized tags. You can just apply the tag-values obtained from the scope in the prometheus handler.

Because the full name including all tags (scope tags and stats tags) is the unique identifier for the stats in the stats allocator. If the scope tags have updated, then the related stats also need to be rebuild. 
Or we can try to support the scople level stats allocator. But this way will introduces some others problems.

**Comment by jmarantz** [comment]
What you suggest would be needed if you have 2 distinct scopes that have the same prefix, and are populated by the same code, and thus share the same stat objects, and then decide while the system is running that you want to drop those shared stats and differentiate them.

If the scopes have distinct prefixes then their stats are already distinct and you don't need to rename them just because you want to visualize more detailed tags for each scope.

Maybe you could describe at a higher level what you are trying to monitor that can't be done currently.

**Comment by jmarantz** [comment]
I'm a little lost as to why all that helps, as well as the impact of a design where you effectively drop a bunch of stat values on a config update, but OK.

Having tags associated with the Scope and inserting them in the Prometheus rendering for each stat in the scope makes sense to me. @ggreenway probably also has opinions about this.

**Comment by wbpcode** [comment]
Yes, this will have some negative effects. However, the data is only lost when the scope tags are updated.
And without doing this, we cannot update the final stats' tags just through xds.

**Comment by wbpcode** [comment]
I also thought of some other things. 

In our practice, we always need to write some very complex regular expressions to extract the cluster name from upstream stats. Because some of our some users' service names are not canonical and require that the non-canonical service names must exist in the final prometheus tags.

If we were able to create a scope with `scopeFromStatNameWithTags` and add the cluster name directly to the scope as scope tags, use cluster.<cluster name> as the scope prefix. Then there is no need to write these complex regular expressions, and it is also compatible with existing stats names.

**Comment by wbpcode** [comment]
The current method of using regular expressions to extract tags is very flexible and powerful.

But if there are some irregular identifiers (such as including variable length dots), the regular expression will be very complicated. And the regular expression for extracting tags can only be configured through bootstrap.

**Comment by wbpcode** [comment]
I just get an idea about tags extraction. It may not a good idea.

We can create scope with prefix and scope tags. The scope tags are only used to populate the final prometheus tags and the prefix used to identify a unique scope. The prefix is consist of one or more segments. Multiple segment values are connected by `.` to form the final prefix. 

And every segment has an optional segment name. The segment name and segment value will be also treated as tags when we try to consume prometheus format stats.

In this way, envoy almost no longer needs to use regular expressions to extract tags. It is also possible to output metrics compatible with the existing statsd/prometheus format. As long as the segment values do not change, we can update the tag name or scope tags arbitrarily through xDS without stats data loss.

**Comment by wbpcode** [comment]
This should replaced by the #20289
